api/v3/Contribution.php

```diff
@@ -54,7 +54,7 @@ function civicrm_api3_contribution_create(&$params) {
       if (empty($params['financial_type_id'])) {
         $params['financial_type_id'] = CRM_Core_DAO::getFieldValue('CRM_Contribute_DAO_Contribution', $params['id'], 'financial_type_id');
       }
-      $op = CRM_Core_Action::UPDATE;
+      $op = CRM_Core_Action::ADD;
```

Comment:
This seems wrong - if id is not set then it must be an UPDATE not an ADD. Are you sure the problem is not in getAvailableFinancialTypes itself?